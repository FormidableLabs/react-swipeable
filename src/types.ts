import * as React from "react";

export const LEFT = "Left";
export const RIGHT = "Right";
export const UP = "Up";
export const DOWN = "Down";
export type HandledEvents = React.MouseEvent | TouchEvent | MouseEvent;
export type Vector2 = [number, number];
export type SwipeDirections =
  | typeof LEFT
  | typeof RIGHT
  | typeof UP
  | typeof DOWN;
export interface SwipeEventData {
  absX: number;
  absY: number;
  deltaX: number;
  deltaY: number;
  dir: SwipeDirections;
  event: HandledEvents;
  first: boolean;
  initial: Vector2;
  velocity: number;
  vxvy: Vector2;
}

export type SwipeCallback = (eventData: SwipeEventData) => void;
export type TapCallback = ({ event }: { event: HandledEvents }) => void;

export type SwipeableCallbacks = {
  // Event handler/callbacks
  onSwipeStart: SwipeCallback;
  onSwiped: SwipeCallback;
  onSwipedDown: SwipeCallback;
  onSwipedLeft: SwipeCallback;
  onSwipedRight: SwipeCallback;
  onSwipedUp: SwipeCallback;
  onSwiping: SwipeCallback;
  onTap: TapCallback;
};

// Configuration Options
export interface ConfigurationOptions {
  delta: number;
  preventDefaultTouchmoveEvent: boolean;
  rotationAngle: number;
  trackMouse: boolean;
  trackTouch: boolean;
}

export type SwipeableProps = Partial<SwipeableCallbacks & ConfigurationOptions>;

export type SwipeablePropsWithDefaultOptions = Partial<SwipeableCallbacks> &
  ConfigurationOptions;

export interface SwipeableHandlers {
  ref(element: HTMLElement | null): void;
  onMouseDown?(event: React.MouseEvent): void;
}

export type SwipeableState = {
  cleanUpTouch?: () => void;
  el?: HTMLElement;
  eventData?: SwipeEventData;
  first: boolean;
  initial: Vector2;
  start: number;
  swiping: boolean;
  xy: Vector2;
};

export type StateSetter = (
  state: SwipeableState,
  props: SwipeablePropsWithDefaultOptions
) => SwipeableState;
export type Setter = (stateSetter: StateSetter) => void;
export type AttachTouch = (el: HTMLElement, passive: boolean) => () => void;
