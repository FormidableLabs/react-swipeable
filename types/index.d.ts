// TypeScript Version: 3.0

import * as React from 'react'

export interface EventData {
  event: MouseEvent | TouchEvent
  deltaX: number
  deltaY: number
  absX: number
  absY: number
  velocity: number
  dir: 'Left' | 'Right' | 'Up' | 'Down'
}

export type SwipeCallback = ({}: EventData) => void

export interface SwipeableOptions {
  // Event handler/callbacks
  onSwiped?: SwipeCallback
  onSwipedLeft?: SwipeCallback
  onSwipedRight?: SwipeCallback
  onSwipedUp?: SwipeCallback
  onSwipedDown?: SwipeCallback
  onSwiping?: SwipeCallback

  // Configuration Props
  delta?: number
  preventDefaultTouchmoveEvent?: boolean
  trackTouch?: boolean
  trackMouse?: boolean
  rotationAngle?: number
}

// Component Specific Props
export interface SwipeableProps {
  nodeName?: string
  innerRef?: ({}: HTMLElement) => void
  children?: React.ReactNode
  style?: React.CSSProperties
  className?: string
}

export class Swipeable extends React.PureComponent<SwipeableProps & SwipeableOptions> {}
