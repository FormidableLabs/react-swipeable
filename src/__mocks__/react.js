/* global window */
const react = require('react');
// Resolution for requestAnimationFrame not supported in jest error :
// https://github.com/facebook/react/issues/9102#issuecomment-283873039
global.window = global;

window.addEventListener = () => {};
window.requestAnimationFrame = () => {
  throw new Error('requestAnimationFrame is not supported in Node');
};

module.exports = react;
