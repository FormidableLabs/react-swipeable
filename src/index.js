/* global document */
import React from 'react'
import PropTypes from 'prop-types'

const defaultProps = {
  preventDefaultTouchmoveEvent: false,
  delta: 10,
  rotationAngle: 0,
  trackMouse: false,
  trackTouch: true
}
const initialState = {
  xy: [0, 0],
  swiping: false,
  lastEventData: undefined,
  start: undefined
}
export const LEFT = 'Left'
export const RIGHT = 'Right'
export const UP = 'Up'
export const DOWN = 'Down'
const touchStart = 'touchstart'
const touchMove = 'touchmove'
const touchEnd = 'touchend'
const mouseMove = 'mousemove'
const mouseUp = 'mouseup'

function getDirection(absX, absY, deltaX, deltaY) {
  if (absX > absY) {
    if (deltaX > 0) {
      return LEFT
    }
    return RIGHT
  } else if (deltaY > 0) {
    return UP
  }
  return DOWN
}

function rotateXYByAngle(pos, angle) {
  if (angle === 0) return pos
  const angleInRadians = (Math.PI / 180) * angle
  const x = pos[0] * Math.cos(angleInRadians) + pos[1] * Math.sin(angleInRadians)
  const y = pos[1] * Math.cos(angleInRadians) - pos[0] * Math.sin(angleInRadians)
  return [x, y]
}

function getHandlers(set, handlerProps) {
  const onStart = event => {
    // if more than a single touch don't track, for now...
    if (event.touches && event.touches.length > 1) return

    set((state, props) => {
      // setup mouse listeners on document to track swipe since swipe can leave container
      if (props.trackMouse) {
        document.addEventListener(mouseMove, onMove)
        document.addEventListener(mouseUp, onUp)
      }
      const { clientX, clientY } = event.touches ? event.touches[0] : event
      const xy = rotateXYByAngle([clientX, clientY], props.rotationAngle)
      return { ...state, ...initialState, xy, start: event.timeStamp || 0 }
    })
  }

  const onMove = event => {
    set((state, props) => {
      if (!state.xy[0] || !state.xy[1] || (event.touches && event.touches.length > 1)) {
        return state
      }
      const { clientX, clientY } = event.touches ? event.touches[0] : event
      const [x, y] = rotateXYByAngle([clientX, clientY], props.rotationAngle)
      const deltaX = state.xy[0] - x
      const deltaY = state.xy[1] - y
      const absX = Math.abs(deltaX)
      const absY = Math.abs(deltaY)
      const time = (event.timeStamp || 0) - state.start
      const velocity = Math.sqrt(absX * absX + absY * absY) / (time || 1)

      // if swipe is under delta and we have not started to track a swipe: skip update
      if (absX < props.delta && absY < props.delta && !state.swiping) return state

      const dir = getDirection(absX, absY, deltaX, deltaY)
      const eventData = { event, absX, absY, deltaX, deltaY, velocity, dir }

      props.onSwiping && props.onSwiping(eventData)

      // track if a swipe is cancelable(handler for swiping or swiped(dir) exists)
      // so we can call preventDefault if needed
      let cancelablePageSwipe = false
      if (props.onSwiping || props.onSwiped || props[`onSwiped${dir}`]) {
        cancelablePageSwipe = true
      }

      if (
        cancelablePageSwipe &&
        props.preventDefaultTouchmoveEvent &&
        props.trackTouch &&
        event.cancelable
      )
        event.preventDefault()

      return { ...state, lastEventData: eventData, swiping: true }
    })
  }

  const onEnd = event => {
    set((state, props) => {
      if (state.swiping) {
        const eventData = { ...state.lastEventData, event }

        props.onSwiped && props.onSwiped(eventData)

        props[`onSwiped${eventData.dir}`] && props[`onSwiped${eventData.dir}`](eventData)
      }
      return { ...state, ...initialState }
    })
  }

  const cleanUpMouse = () => {
    // safe to just call removeEventListener
    document.removeEventListener(mouseMove, onMove)
    document.removeEventListener(mouseUp, onUp)
  }

  const onUp = e => {
    cleanUpMouse()
    onEnd(e)
  }

  const attachTouch = el => {
    if (el && el.addEventListener) {
      // attach touch event listeners and handlers
      const tls = [[touchStart, onStart], [touchMove, onMove], [touchEnd, onEnd]]
      tls.forEach(([e, h]) => el.addEventListener(e, h))
      // return properly scoped cleanup method for removing listeners
      return () => tls.forEach(([e, h]) => el.removeEventListener(e, h))
    }
  }

  const onRef = el => {
    // "inline" ref functions are called twice on render, once with null then again with DOM element
    // ignore null here
    if (el === null) return
    set((state, props) => {
      // if the same DOM el as previous just return state
      if (state.el === el) return state

      let addState = {}
      // if new DOM el clean up old DOM and reset cleanUpTouch
      if (state.el && state.el !== el && state.cleanUpTouch) {
        state.cleanUpTouch()
        addState.cleanUpTouch = null
      }
      // only attach if we want to track touch
      if (props.trackTouch && el) {
        addState.cleanUpTouch = attachTouch(el)
      }

      // store event attached DOM el for comparison, clean up, and re-attachment
      return { ...state, el, ...addState }
    })
  }

  // update state, and handlers
  set((state, props) => {
    let addState = {}
    // clean up touch handlers if no longer tracking touches
    if (!props.trackTouch && state.cleanUpTouch) {
      state.cleanUpTouch()
      addState.cleanUpTouch = null
    } else if (props.trackTouch && !state.cleanUpTouch) {
      // attach/re-attach touch handlers
      if (state.el) {
        addState.cleanUpTouch = attachTouch(state.el)
      }
    }
    return { ...state, ...addState }
  })

  // set ref callback to attach touch event listeners
  const output = { ref: onRef }

  // if track mouse attach mouse down listener
  if (handlerProps.trackMouse) {
    output.onMouseDown = onStart
  }

  return output
}

export function useSwipeable(props) {
  const transientState = React.useRef({ ...initialState, type: 'hook' })
  const transientProps = React.useRef()
  transientProps.current = { ...defaultProps, ...props }
  return getHandlers(
    cb => (transientState.current = cb(transientState.current, transientProps.current)),
    { trackMouse: props.trackMouse }
  )
}

export class Swipeable extends React.PureComponent {
  static propTypes = {
    onSwiped: PropTypes.func,
    onSwiping: PropTypes.func,
    onSwipedUp: PropTypes.func,
    onSwipedRight: PropTypes.func,
    onSwipedDown: PropTypes.func,
    onSwipedLeft: PropTypes.func,
    delta: PropTypes.number,
    preventDefaultTouchmoveEvent: PropTypes.bool,
    nodeName: PropTypes.string,
    trackMouse: PropTypes.bool,
    trackTouch: PropTypes.bool,
    innerRef: PropTypes.func,
    rotationAngle: PropTypes.number
  }

  static defaultProps = defaultProps

  constructor(props) {
    super(props)
    this.transientState = { ...initialState, type: 'class' }
  }

  _set = cb => {
    this.transientState = cb(this.transientState, this.props)
  }

  render() {
    const { className, style, nodeName = 'div', innerRef, children, trackMouse } = this.props
    const handlers = getHandlers(this._set, { trackMouse })
    const ref = innerRef ? el => (innerRef(el), handlers.ref(el)) : handlers.ref
    return React.createElement(nodeName, { ...handlers, className, style, ref }, children)
  }
}
