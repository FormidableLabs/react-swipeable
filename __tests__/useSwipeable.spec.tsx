import * as React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import { useSwipeable } from "../src/index";
import { LEFT, RIGHT, UP, DOWN } from "../src/types";
import { expectSwipeFuncsDir } from "./helpers";

const DIRECTIONS: [typeof LEFT, typeof RIGHT, typeof UP, typeof DOWN] = [
  LEFT,
  RIGHT,
  UP,
  DOWN,
];

export type MockedSwipeFunctions = {
  onSwiping: jest.Mock;
  onSwiped: jest.Mock;
  onSwipedLeft: jest.Mock;
  onSwipedRight: jest.Mock;
  onSwipedUp: jest.Mock;
  onSwipedDown: jest.Mock;
};
function getMockedSwipeFunctions(): MockedSwipeFunctions {
  const onSwiped = "onSwiped";
  return DIRECTIONS.reduce(
    (acc, dir) => ({ ...acc, [onSwiped + dir]: jest.fn() }),
    {
      onSwiping: jest.fn(),
      onSwiped: jest.fn(),
    } as MockedSwipeFunctions
  );
}

const TESTING_TEXT = "touch here";
/*
 * Wrapping component for the hook testing
 */
function SwipeableUsingHook({ nodeName = "div", ...rest }) {
  const eventHandlers = useSwipeable(rest);
  const Elem = nodeName as React.ElementType;
  return (
    <Elem {...eventHandlers}>
      <span>{TESTING_TEXT}</span>
    </Elem>
  );
}

const MD = "mouseDown";
const MM = "mouseMove";
const MU = "mouseUp";
const TS = "touchStart";
const TM = "touchMove";
const TE = "touchEnd";

const createClientXYObject = (x?: number, y?: number) => ({
  clientX: x,
  clientY: y,
});
// Create touch event
const cte = ({ x, y }: { x?: number; y?: number }) => ({
  touches: [createClientXYObject(x, y)],
});
// Create Mouse Event
const cme = ({ x, y }: { x?: number; y?: number }) => ({
  ...createClientXYObject(x, y),
});

describe("useSwipeable", () => {
  let defaultPrevented = 0;

  beforeAll(() => {
    // listen on document for touchmove events, track if their default was prevented
    document.addEventListener("touchmove", (e) => {
      if (e.defaultPrevented) {
        defaultPrevented += 1;
      }
    });
  });

  beforeEach(() => {
    defaultPrevented = 0;
  });

  it("handles onTap callbacks", () => {
    const onTap = jest.fn();
    const swipeFuncs = getMockedSwipeFunctions();
    const { getByText } = render(
      <SwipeableUsingHook {...swipeFuncs} onTap={onTap} />
    );
    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TE](touchArea, cte({ x: 100, y: 100 }));

    expect(onTap).toHaveBeenCalled();
  });

  it("handles touch events that start at clientX or clientY 0", () => {
    const swipeFuncs = getMockedSwipeFunctions();
    const { getByText } = render(<SwipeableUsingHook {...swipeFuncs} />);

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 0, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 0, y: 125 }));
    fireEvent[TE](touchArea, cte({}));
    fireEvent[TS](touchArea, cte({ x: 100, y: 0 }));
    fireEvent[TM](touchArea, cte({ x: 125, y: 0 }));
    fireEvent[TE](touchArea, cte({}));

    expect(swipeFuncs.onSwiping).toHaveBeenCalledTimes(2);
    expect(swipeFuncs.onSwiped).toHaveBeenCalledTimes(2);
  });

  it("handles touch events and fires correct props", () => {
    const onTap = jest.fn();
    const swipeFuncs = getMockedSwipeFunctions();
    const { getByText } = render(
      <SwipeableUsingHook {...swipeFuncs} onTap={onTap} />
    );

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 125 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 150 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 175 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 200 }));
    fireEvent[TE](touchArea, cte({}));

    expect(swipeFuncs.onSwipedDown).toHaveBeenCalled();
    expect(swipeFuncs.onSwipedUp).not.toHaveBeenCalled();
    expect(swipeFuncs.onSwipedLeft).not.toHaveBeenCalled();
    expect(swipeFuncs.onSwipedRight).not.toHaveBeenCalled();
    expect(swipeFuncs.onSwiped.mock.calls[0][0]).toMatchSnapshot(
      {
        velocity: expect.any(Number),
        vxvy: expect.arrayContaining([expect.any(Number), expect.any(Number)]),
      },
      `useSwipeable onSwiped trackTouch`
    );
    expect(swipeFuncs.onSwiping.mock.calls[0][0]).toMatchSnapshot(
      {
        velocity: expect.any(Number),
        vxvy: expect.arrayContaining([expect.any(Number), expect.any(Number)]),
      },
      `useSwipeable onSwiping trackTouch`
    );
    expect(onTap).not.toHaveBeenCalled();
  });

  it("handles mouse events with trackMouse prop and fires correct props", () => {
    const swipeFuncs = getMockedSwipeFunctions();
    const { getByText } = render(
      <SwipeableUsingHook
        {...swipeFuncs}
        trackMouse={true}
        trackTouch={false}
      />
    );

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[MD](touchArea, cme({ x: 100, y: 100 }));
    fireEvent[MM](touchArea, cme({ x: 125, y: 100 }));
    fireEvent[MM](touchArea, cme({ x: 150, y: 100 }));
    fireEvent[MM](touchArea, cme({ x: 175, y: 100 }));
    fireEvent[MM](touchArea, cme({ x: 200, y: 100 }));
    fireEvent[MU](document, cme({}));

    expect(swipeFuncs.onSwipedRight).toHaveBeenCalled();
    expect(swipeFuncs.onSwipedUp).not.toHaveBeenCalled();
    expect(swipeFuncs.onSwipedDown).not.toHaveBeenCalled();
    expect(swipeFuncs.onSwipedLeft).not.toHaveBeenCalled();
    expect(swipeFuncs.onSwiped.mock.calls[0][0]).toMatchSnapshot(
      { velocity: expect.any(Number), vxvy: expect.any(Array) },
      `useSwipeable onSwiped trackMouse`
    );
    expect(swipeFuncs.onSwiping.mock.calls[0][0]).toMatchSnapshot(
      { velocity: expect.any(Number), vxvy: expect.any(Array) },
      `useSwipeable onSwiping trackMouse`
    );
  });

  it("calls preventDefault when swiping in direction that has a callback", () => {
    const onSwipedDown = jest.fn();

    const { getByText } = render(
      <SwipeableUsingHook
        onSwipedDown={onSwipedDown}
        preventDefaultTouchmoveEvent
      />
    );

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 125 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 150 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 175 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 200 }));
    fireEvent[TE](touchArea, cte({}));

    expect(onSwipedDown).toHaveBeenCalled();
    expect(defaultPrevented).toBe(4);
  });

  it("does not call preventDefault when false", () => {
    const onSwipedUp = jest.fn();

    const { getByText } = render(
      <SwipeableUsingHook onSwipedUp={onSwipedUp} />
    );

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 75 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 50 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 25 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 5 }));
    fireEvent[TE](touchArea, cte({}));

    expect(onSwipedUp).toHaveBeenCalled();
    expect(defaultPrevented).toBe(0);
  });

  it("calls preventDefault when onSwiping or onSwiped is present", () => {
    const onSwiping = jest.fn();
    const onSwiped = jest.fn();

    const { getByText, rerender } = render(
      <SwipeableUsingHook onSwiping={onSwiping} preventDefaultTouchmoveEvent />
    );

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 50 }));
    fireEvent[TE](touchArea, cte({}));

    expect(onSwiping).toHaveBeenCalled();
    expect(defaultPrevented).toBe(1);

    rerender(
      <SwipeableUsingHook onSwiped={onSwiped} preventDefaultTouchmoveEvent />
    );

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 50 }));
    fireEvent[TE](touchArea, cte({}));

    expect(onSwiped).toHaveBeenCalled();
    expect(defaultPrevented).toBe(2);
  });

  it("does not re-check delta when swiping already in progress", () => {
    const onSwiping = jest.fn();
    const onSwipedRight = jest.fn();
    const onSwipedLeft = jest.fn();
    const { getByText } = render(
      <SwipeableUsingHook
        onSwiping={onSwiping}
        onSwipedRight={onSwipedRight}
        onSwipedLeft={onSwipedLeft}
        delta={40}
      />
    );

    const touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 145, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 80, y: 100 }));
    fireEvent[TE](touchArea, cte({}));

    expect(onSwiping).toHaveBeenCalledTimes(2);
    expect(onSwiping.mock.calls[0][0].dir).toBe(RIGHT);
    expect(onSwiping.mock.calls[1][0].dir).toBe(LEFT);
    expect(onSwipedLeft).toHaveBeenCalledTimes(1);
    expect(onSwipedRight).not.toHaveBeenCalled();
  });

  it("Handle Rotation by 90 degree", () => {
    const swipeFuncsRight = getMockedSwipeFunctions();
    const { getByText, rerender } = render(
      <SwipeableUsingHook {...swipeFuncsRight} rotationAngle={90} />
    );

    const touchArea = getByText(TESTING_TEXT);

    // check right
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 125 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 150 }));
    fireEvent[TE](touchArea, cte({}));
    expectSwipeFuncsDir(swipeFuncsRight, RIGHT);

    // check left
    const swipeFuncsLeft = getMockedSwipeFunctions();
    rerender(<SwipeableUsingHook {...swipeFuncsLeft} rotationAngle={90} />);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 75 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 50 }));
    fireEvent[TE](touchArea, cte({}));
    expectSwipeFuncsDir(swipeFuncsLeft, LEFT);

    // check up
    const swipeFunsUp = getMockedSwipeFunctions();
    rerender(<SwipeableUsingHook {...swipeFunsUp} rotationAngle={90} />);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 125, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 150, y: 100 }));
    fireEvent[TE](touchArea, cte({}));
    expectSwipeFuncsDir(swipeFunsUp, UP);

    // check down
    const swipeFunsDown = getMockedSwipeFunctions();
    rerender(<SwipeableUsingHook {...swipeFunsDown} rotationAngle={90} />);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 75, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 50, y: 100 }));
    fireEvent[TE](touchArea, cte({}));
    expectSwipeFuncsDir(swipeFunsDown, DOWN);
  });

  it('Handle "odd" rotations', () => {
    const swipeFuncsNegativeRotation = getMockedSwipeFunctions();
    const { getByText, rerender } = render(
      <SwipeableUsingHook {...swipeFuncsNegativeRotation} rotationAngle={-90} />
    );

    const touchArea = getByText(TESTING_TEXT);

    // check -90
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 125 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 150 }));
    fireEvent[TE](touchArea, cte({}));
    expectSwipeFuncsDir(swipeFuncsNegativeRotation, LEFT);

    // check 360 + 270
    const swipeFuncsLargeRotation = getMockedSwipeFunctions();
    rerender(
      <SwipeableUsingHook
        {...swipeFuncsLargeRotation}
        rotationAngle={360 + 270}
      />
    );
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 125 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 150 }));
    fireEvent[TE](touchArea, cte({}));
    expectSwipeFuncsDir(swipeFuncsLargeRotation, LEFT);
  });

  it("Handle Rotation that changes so keep the direction the same", () => {
    const swipeFuncs = getMockedSwipeFunctions();
    const { getByText, rerender } = render(
      <SwipeableUsingHook {...swipeFuncs} />
    );

    const touchArea = getByText(TESTING_TEXT);

    // check 0
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 125, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 150, y: 100 }));
    fireEvent[TE](touchArea, cte({}));
    expect(swipeFuncs.onSwiped).toHaveBeenCalledTimes(1);
    expect(swipeFuncs.onSwipedRight).toHaveBeenCalledTimes(1);

    // check 90
    rerender(<SwipeableUsingHook {...swipeFuncs} rotationAngle={90} />);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 125 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 150 }));
    fireEvent[TE](touchArea, cte({}));
    expect(swipeFuncs.onSwiped).toHaveBeenCalledTimes(2);
    expect(swipeFuncs.onSwipedRight).toHaveBeenCalledTimes(2);

    // check 180
    rerender(<SwipeableUsingHook {...swipeFuncs} rotationAngle={180} />);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 75, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 50, y: 100 }));
    fireEvent[TE](touchArea, cte({}));
    expect(swipeFuncs.onSwiped).toHaveBeenCalledTimes(3);
    expect(swipeFuncs.onSwipedRight).toHaveBeenCalledTimes(3);

    // check 270
    rerender(<SwipeableUsingHook {...swipeFuncs} rotationAngle={270} />);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 75 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 50 }));
    fireEvent[TE](touchArea, cte({}));
    expect(swipeFuncs.onSwiped).toHaveBeenCalledTimes(4);
    expect(swipeFuncs.onSwipedRight).toHaveBeenCalledTimes(4);

    expect(swipeFuncs.onSwiping).toHaveBeenCalledTimes(8);
    [LEFT, UP, DOWN].forEach((dir) => {
      expect(
        swipeFuncs[`onSwiped${dir}` as keyof MockedSwipeFunctions]
      ).not.toHaveBeenCalled();
    });
  });

  it("Does not track touches when trackTouch is false", () => {
    const onSwipedDown = jest.fn();

    const { getByText, rerender } = render(
      <SwipeableUsingHook onSwipedDown={onSwipedDown} />
    );

    const touchArea = getByText(TESTING_TEXT);

    const doSwipe = () => {
      fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
      fireEvent[TM](touchArea, cte({ x: 100, y: 200 }));
      fireEvent[TE](touchArea, cte({}));
    };

    doSwipe();
    expect(onSwipedDown).toHaveBeenCalledTimes(1);

    rerender(
      <SwipeableUsingHook onSwipedDown={onSwipedDown} trackTouch={false} />
    );
    doSwipe();
    // verify we did not trigger another swipe
    expect(onSwipedDown).toHaveBeenCalledTimes(1);

    rerender(
      <SwipeableUsingHook onSwipedDown={onSwipedDown} trackTouch={true} />
    );
    doSwipe();
    expect(onSwipedDown).toHaveBeenCalledTimes(2);
  });

  it("Cleans up and re-attaches touch event listeners if the DOM element changes", () => {
    const onSwipedDown = jest.fn();

    const { getByText, rerender } = render(
      <SwipeableUsingHook onSwipedDown={onSwipedDown} />
    );

    let touchArea = getByText(TESTING_TEXT);

    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 200 }));
    fireEvent[TE](touchArea, cte({}));

    expect(onSwipedDown).toHaveBeenCalledTimes(1);

    rerender(<SwipeableUsingHook onSwipedDown={onSwipedDown} nodeName="p" />);
    // re-get element since wrapping dom node changed
    touchArea = getByText(TESTING_TEXT);
    fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
    fireEvent[TM](touchArea, cte({ x: 100, y: 200 }));
    fireEvent[TE](touchArea, cte({}));
    // verify we triggered callbacks on new dom element
    expect(onSwipedDown).toHaveBeenCalledTimes(2);
  });

  it(`handles new prop swipe callbacks from re-renders`, () => {
    const onSwipedSpy = jest.fn();

    function TestHookComponent({ next }: { next: () => void }) {
      const handlers = useSwipeable({ onSwiped: next });
      return <div {...handlers}>{TESTING_TEXT}</div>;
    }

    function TestComponent() {
      const [page, setPage] = React.useState(0);
      // Changing the callback on each re-render ON PURPOSE
      // In order to validate useSwipeable updates prop references
      const next = () => (setPage(page + 1), onSwipedSpy(page + 1));

      return <TestHookComponent next={next} />;
    }

    const { getByText } = render(<TestComponent />);

    const touchArea = getByText(TESTING_TEXT);

    const pages = [1, 2, 3];
    // swipe left 3 times
    pages.forEach(() => {
      act(() => {
        fireEvent[TS](touchArea, cte({ x: 100, y: 100 }));
        fireEvent[TM](touchArea, cte({ x: 75, y: 100 }));
        fireEvent[TE](touchArea, cte({}));
      });
    });
    pages.forEach((page, idx) => {
      expect(onSwipedSpy.mock.calls[idx][0]).toBe(page);
    });
  });
});
