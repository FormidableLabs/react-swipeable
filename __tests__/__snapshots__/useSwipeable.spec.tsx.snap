// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSwipeable handles mouse events with trackMouse prop and fires correct props: useSwipeable onSwiped trackMouse 1`] = `
Object {
  "absX": 100,
  "absY": 0,
  "deltaX": 100,
  "deltaY": 0,
  "dir": "Right",
  "event": MouseEvent {
    "isTrusted": false,
  },
  "first": false,
  "initial": Array [
    100,
    100,
  ],
  "velocity": Any<Number>,
  "vxvy": Any<Array>,
}
`;

exports[`useSwipeable handles mouse events with trackMouse prop and fires correct props: useSwipeable onSwiping trackMouse 1`] = `
Object {
  "absX": 25,
  "absY": 0,
  "deltaX": 25,
  "deltaY": 0,
  "dir": "Right",
  "event": MouseEvent {
    "isTrusted": false,
  },
  "first": true,
  "initial": Array [
    100,
    100,
  ],
  "velocity": Any<Number>,
  "vxvy": Any<Array>,
}
`;

exports[`useSwipeable handles touch events and fires correct props: useSwipeable onSwiped trackTouch 1`] = `
Object {
  "absX": 0,
  "absY": 100,
  "deltaX": 0,
  "deltaY": 100,
  "dir": "Down",
  "event": TouchEvent {
    "isTrusted": false,
  },
  "first": false,
  "initial": Array [
    100,
    100,
  ],
  "velocity": Any<Number>,
  "vxvy": ArrayContaining [
    Any<Number>,
    Any<Number>,
  ],
}
`;

exports[`useSwipeable handles touch events and fires correct props: useSwipeable onSwiping trackTouch 1`] = `
Object {
  "absX": 0,
  "absY": 25,
  "deltaX": 0,
  "deltaY": 25,
  "dir": "Down",
  "event": TouchEvent {
    "isTrusted": false,
  },
  "first": true,
  "initial": Array [
    100,
    100,
  ],
  "velocity": Any<Number>,
  "vxvy": ArrayContaining [
    Any<Number>,
    Any<Number>,
  ],
}
`;
